nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/ADLSSecureCredentialsControllerService.java

```diff
@@ -0,0 +1,157 @@
+    ));
+
+    private ConfigurationContext context;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> results = new ArrayList<>();
+
+        boolean accountKeySet = StringUtils.isNotBlank(validationContext.getProperty(
+                AzureStorageUtils.ACCOUNT_KEY_SECRET).getValue());
+        boolean sasTokenSet = StringUtils.isNotBlank(validationContext.getProperty(
+                AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET).getValue());
+
+        if (!onlyOneSet(accountKeySet, sasTokenSet)) {
+            StringJoiner options = new StringJoiner(", ")
+                    .add(AzureStorageUtils.ACCOUNT_KEY_SECRET.getDisplayName())
+                    .add(AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET.getDisplayName());
+
+            results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
+                    .valid(false)
+                    .explanation("one and only one of [" + options + "] should be set")
+                    .build());
+        }
+
+        return results;
+    }
+
+    private boolean onlyOneSet(Boolean... checks) {
+        long nrOfSet = Arrays.stream(checks)
+                .filter(check -> check)
+                .count();
+
+        return nrOfSet == 1;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public ADLSCredentialsDetails getCredentialsDetails(Map<String, String> attributes) {
+        ADLSCredentialsDetails.Builder credentialsBuilder = ADLSCredentialsDetails.Builder.newBuilder();
+
+        setValue(credentialsBuilder,
+                AzureStorageUtils.ADLS_ENDPOINT_SUFFIX,
+                PropertyValue::getValue,
+                ADLSCredentialsDetails.Builder::setEndpointSuffix);
+
+        final AzureKeyVaultConnectionService keyVaultClientService = context.getProperty(
+                AzureStorageUtils.KEYVAULT_CONNECTION_SERVICE
+        ).asControllerService(AzureKeyVaultConnectionService.class);
+        String accountNameSecret = context.getProperty(
+                AzureStorageUtils.ACCOUNT_NAME_SECRET).getValue();
+        String accountKeySecret = context.getProperty(
+                AzureStorageUtils.ACCOUNT_KEY_SECRET).getValue();
+        String sasTokenSecret = context.getProperty(
+                AzureStorageUtils.ACCOUNT_SAS_TOKEN_SECRET).getValue();
+
+        if(StringUtils.isNotBlank(accountNameSecret)) {
+            credentialsBuilder.setAccountName(
+                    keyVaultClientService.getSecret(accountNameSecret));
+        }
+        if(StringUtils.isNotBlank(accountKeySecret)) {
+            credentialsBuilder.setAccountKey(
+                    keyVaultClientService.getSecret(accountKeySecret));
```

Comment:
what if secret is not set for this key in keyvault? or key is mispelled? Does this throw exception?

Reply:
Keyvault does not allows empty secret, in case of a non-existent secret requested there would be an error/exception from keyvault sdk